src/coreComponents/common/logger/ErrorHandling.cpp

```diff
@@ -0,0 +1,197 @@
+  }
+}
+
+void ErrorLogger::ErrorMsg::addContextInfoImpl( ErrorLogger::ContextInfo && ctxInfo )
+{
+  m_contextsInfo.emplace_back( std::move( ctxInfo ) );
+}
+
+ErrorLogger::ErrorMsg & ErrorLogger::ErrorMsg::setRank( int rank )
+{
+  m_ranksInfo.push_back( rank );
+  return *this;
+}
+
+ErrorLogger::ErrorMsg & ErrorLogger::ErrorMsg::addCallStackInfo( std::string const & ossStackTrace )
+{
+  std::istringstream iss( ossStackTrace );
+  std::string stackLine;
+  std::size_t index;
+
+  while( std::getline( iss, stackLine ) )
+  {
+    index = stackLine.find( ':' );
+    m_sourceCallStack.push_back( stackLine.substr( index + 1 ) );
+  }
+
+  return *this;
+}
+
+std::string ErrorLogger::toString( ErrorLogger::MsgType type )
+{
+  switch( type )
+  {
+    case ErrorLogger::MsgType::Error: return "Error";
+    case ErrorLogger::MsgType::Warning: return "Warning";
+    case ErrorLogger::MsgType::Exception: return "Exception";
+    default: return "Unknown";
+  }
+}
+
+ErrorLogger::ErrorMsg & ErrorLogger::ErrorMsg::addToMsg( std::exception const & e )
+{
+  parent->m_currentErrorMsg.m_msg = e.what();
+  return parent->m_currentErrorMsg;
+}
+
+ErrorLogger::ErrorMsg & ErrorLogger::ErrorMsg::addToMsg( std::string const & errorMsg )
+{
+  parent->m_currentErrorMsg.m_msg = errorMsg + parent->m_currentErrorMsg.m_msg; 
+  return parent->m_currentErrorMsg;
+}
+
+ErrorLogger::ErrorMsg & ErrorLogger::ErrorMsg::setCodeLocation( string msgFile, integer msgLine )
+{
+  parent->m_currentErrorMsg.m_file = msgFile;
+  parent->m_currentErrorMsg.m_line = msgLine;
+  return parent->m_currentErrorMsg;
+}
+
+ErrorLogger::ErrorMsg & ErrorLogger::ErrorMsg::setType( ErrorLogger::MsgType msgType )
+{
+  parent->m_currentErrorMsg.m_type = msgType;
+  return parent->m_currentErrorMsg;
+}
+
+void ErrorLogger::streamMultilineYamlAttribute( std::string_view msg, std::ofstream& yamlFile )
+{
+  while( !msg.empty() )
+  {
+    const size_t index = msg.find( "\n" );
+    std::string_view line = msg.substr( 0, index );
+    yamlFile << g_level2Next << line << "\n";
+
+    if( index == msg.npos )
+      break;
+    msg.remove_prefix( index + 1 );
+  }
+}
+
+void ErrorLogger::write( ErrorLogger::ErrorMsg const & errorMsg ) //const
+{
+  std::ofstream yamlFile( std::string( m_filename ), std::ios::app );
+  if( yamlFile.is_open() )
+  {
+    yamlFile << "\n" << g_level1Start << "type: " << errorLogger.toString( errorMsg.m_type ) << "\n";
+    yamlFile << g_level1Next << "rank: ";
+    for( size_t i = 0; i < errorMsg.m_ranksInfo.size(); i++ )
+    {
+      yamlFile << errorMsg.m_ranksInfo[i];
+    }
+    yamlFile << "\n";
+    yamlFile << g_level1Next << "message: >-\n";
+    streamMultilineYamlAttribute( errorMsg.m_msg, yamlFile );
+    if( !errorMsg.m_contextsInfo.empty() )
+    {
+      yamlFile << g_level1Next << "contexts:\n";
+      for( size_t i = 0; i < errorMsg.m_contextsInfo.size(); i++ )
+      {
```

Comment:
If you don't really use the id, prefer a for-each syntax:
```suggestion
      for( ErrorMsg const & ctxInfo : errorMsg.m_contextsInfo.size() )
      {
```
- less prone to confusion & bugs when staking for-loops,
- better shows your intention, clearer code,
- strong typing is optional, but can improve clarity & reliability,
- you can specify if your loop is for read-only (`const`),
- can be faster in a few cases.

Reply:
I think there still are some

Reply:
It is done. I replaced, when necessary, the for loops with id by for-each.